### PR TITLE
Rm superuser require all spaces

### PR DESCRIPTION
**Pull request by XavierM**
to fix : https://github.com/elastic/kibana/pull/118001#issuecomment-984131327

### CODE REVIEW

File: x-pack/plugins/fleet/server/plugin.ts
Comment (jportner) [inline x-pack/plugins/fleet/server/plugin.ts]: Don't forget to remove the changes in this file before merging!

File: x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.tsx
Comment (jportner) [inline x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.tsx]: Is it realistic that the UI tries to find a featureId that is not defined?
Maybe this would be easier to reason about and test:

```suggestion
      if (featureId) {
        securedFeaturesToSet = [securedFeaturesToSet.find((sf) => sf.id === featureId)!];
      }
```

File: x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.tsx
Comment (jportner) [inline x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.tsx]: I haven't tested this but reading the logic, I think this might be a more accurate conditional:

```suggestion
          if (
            !pfp.disabled &&
            (!pfp.requireAllSpaces || this.state.selectedSpaceIds.includes(ALL_SPACES_ID))
          ) {
            return privileges.includes(pfp.id);
          }
```

WDYT?

**Review by jportner — COMMENTED** [review]
C'est magnifique!

**Comment by criamico** [comment]
I tested locally and works as expected! Thanks Xavier :)

File: x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/privilege_form_calculator/privilege_form_calculator.ts
Comment (legrego) [inline x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/privilege_form_calculator/privilege_form_calculator.ts:95]: nit: missing jsdocs for new `allSpacesSelected` parameter

```suggestion
   * @param privilegeIndex the index of the kibana privileges role component
   * @param allSpacesSelected indicates if the privilege form is configured to grant access to all spaces.
```

File: x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/privilege_form_calculator/privilege_form_calculator.ts
Comment (legrego) [inline x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/privilege_form_calculator/privilege_form_calculator.ts]: The phrase "displayed privilege" means something different, so we should probably rename this to avoid confusion:

```suggestion
    const effectivePrivilege = feature
      .getPrimaryFeaturePrivileges({ includeMinimalFeaturePrivileges: true })
      .find((fp) => {
        return selectedFeaturePrivileges.includes(fp.id) || basePrivilege?.grantsPrivilege(fp);
      });
    const correctSpacesSelected = effectivePrivilege?.requireAllSpaces ? allSpacesSelected : true;
    const availablePrivileges = correctSpacesSelected && !effectivePrivilege?.disabled;
    if (availablePrivileges) return effectivePrivilege;
```

File: x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.test.tsx
Comment (legrego) [inline x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.test.tsx]:
```suggestion
  describe('Feature with a disabled `read` privilege', () => {
```

File: x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.test.tsx
Comment (legrego) [inline x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.test.tsx]: I know we talked about this yesterday, but now that I see it in person, I don't think this asserts anything new. The `expect` call below should be asserting the same thing based on how the feature privileges are defined there.

If you think this improves the readability/intentionality of the test, then I'm fine to leave it in. Your call!

File: x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.tsx
Comment (legrego) [inline x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.tsx]: If the feature does not have any associated privileges, then we shouldn't be creating an entry in the role definition for it. What do you think about something like this?

```suggestion
  private resetRoleFeature = (roleFeature: FeaturesPrivileges) => {
    const securedFeatures = this.props.kibanaPrivileges.getSecuredFeatures();
    return Object.entries(roleFeature).reduce((features, [featureId, privileges]) => {
      if (!Array.isArray(privileges)) {
        return features;
      }
      const securedFeature = securedFeatures.find((sf) => sf.id === featureId);
      const primaryFeaturePrivilege = securedFeature
        ?.getPrimaryFeaturePrivileges({ includeMinimalFeaturePrivileges: true })
        .find((pfp) => {
          if (
            pfp?.disabled ||
            (pfp?.requireAllSpaces && !this.state.selectedSpaceIds.includes(ALL_SPACES_ID))
          ) {
            return false;
          }
          return privileges.includes(pfp.id);
        }) ?? { disabled: false, requireAllSpaces: false };
      return {
        ...features,
        [featureId]: privileges.filter((p) => {
          if (
            primaryFeaturePrivilege?.disabled ||
            (primaryFeaturePrivilege?.requireAllSpaces &&
              !this.state.selectedSpaceIds.includes(ALL_SPACES_ID))
          ) {
            return false;
          }
          return true;
        }),
      };
    }, {});
  };
```

File: x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.tsx
Comment (legrego) [inline x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.tsx:548]: This still suffers from the problem we identified yesterday, where we aren't unsetting the `All` privilege for Fleet when bulk-changing from `All` to `Read`:

1) Configure Fleet to `requireAllSpaces` for the `all` privilege
2) Disable Fleet's `read` privilege
3) Go to role management, start to create a new role
4) Add Kibana privilege for All Spaces
5) Bulk change to `All`
6) Bulk change to `Read`

Expect: Fleet changes from `All` to `None`
Actual: Fleet stays at `All`

File: x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.test.tsx
Comment (XavierM) [reply x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.test.tsx]: I think it make sense to do it when we are checking if the feature is undefined, I think it is more obvious. However for the other ones I agree with you

File: x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.tsx
Comment (XavierM) [reply x-pack/plugins/security/public/management/roles/edit_role/privileges/kibana/space_aware_privilege_section/privilege_space_form.tsx:548]: Yes, I messed up here, I tried to do something cleaner and I got it wrong. I just wanted to make sure that you saw it too, nope!

**Review by legrego — APPROVED** [review]

**Pull request merged**
